### PR TITLE
Docker-compose backup and upgrade commands

### PR DESCRIPTION
**Pull request by Gaitholabi**

### CODE REVIEW

File: templates/docker-compose/Makefile [hunk 1]
```diff
@@ -73,3 +74,17 @@ configure:
 	cp -n docker-compose.yml.dist docker-compose.yml || true
 	cp -n docker-compose.prod.yml.dist docker-compose.prod.yml || true
 	cp -n -R ../../configs/gateway . || true
+
+.PHONY: upgrade
+upgrade: backup
```
Comment (bytestream) [inline templates/docker-compose/Makefile:79]:
```suggestion
upgrade: backup ## Upgrade SupportPal to a later version.
```

File: templates/docker-compose/Makefile [hunk 2]
```diff
@@ -73,3 +74,17 @@ configure:
...
+	$(DOCKER_COMPOSE_BIN) $(COMPOSE_FILES) down -v
+	$(MAKE) start
+
+.PHONY: backup
+backup:
```
Comment (bytestream) [inline templates/docker-compose/Makefile:84]:
```suggestion
backup: ## Create a full system backup.
```

File: templates/docker-compose/Makefile [hunk 3]
```diff
@@ -1,4 +1,5 @@
 SELF_FILENAME := $(word $(words $(MAKEFILE_LIST)),$(MAKEFILE_LIST))
+TIMESTAMP := $$(date +%s)
```
Comment (jshah4517) [inline templates/docker-compose/Makefile:2]: What does this produce? `Y-m-d` kind of format would be good.

File: templates/docker-compose/Makefile [hunk 4]
```diff
@@ -73,3 +74,17 @@ configure:
...
 	cp -n -R ../../configs/gateway . || true
+
+.PHONY: upgrade
+upgrade: backup
+	$(DOCKER_COMPOSE_BIN) $(COMPOSE_FILES) down -v
+	$(MAKE) start
+
+.PHONY: backup
+backup:
+	mkdir -p "backup/$(TIMESTAMP)/mysql/" && docker cp "$(DATABASE_SERVICE_NAME):/var/lib/mysql" "backup/$(TIMESTAMP)/mysql/"
```
Comment (bytestream) [inline templates/docker-compose/Makefile:85]:
```suggestion
	@ mkdir -p "backup/$(TIMESTAMP)/mysql/" && docker cp "$(DATABASE_SERVICE_NAME):/var/lib/mysql" "backup/$(TIMESTAMP)/mysql/"
```

File: templates/docker-compose/Makefile [hunk 5]
```diff
@@ -73,3 +74,17 @@ configure:
...
+	mkdir -p "backup/$(TIMESTAMP)/cache/" && docker cp "$(CACHE_SERVICE_NAME):/data" "backup/$(TIMESTAMP)/cache/"
```
Comment (bytestream) [inline templates/docker-compose/Makefile:86]:
```suggestion
	@ mkdir -p "backup/$(TIMESTAMP)/cache/" && docker cp "$(CACHE_SERVICE_NAME):/data" "backup/$(TIMESTAMP)/cache/"
```

File: templates/docker-compose/Makefile [hunk 6]
```diff
@@ -73,3 +74,17 @@ configure:
...
+	mkdir -p "backup/$(TIMESTAMP)/mailer/" && docker cp "$(MAILER_SERVICE_NAME):/var/spool/exim4" "backup/$(TIMESTAMP)/mailer/"
```
Comment (bytestream) [inline templates/docker-compose/Makefile:87]:
```suggestion
	@ mkdir -p "backup/$(TIMESTAMP)/mailer/" && docker cp "$(MAILER_SERVICE_NAME):/var/spool/exim4" "backup/$(TIMESTAMP)/mailer/"
```

File: templates/docker-compose/Makefile [hunk 7]
```diff
@@ -73,3 +74,17 @@ configure:
...
+	mkdir -p "backup/$(TIMESTAMP)/config/" && docker cp "$(WEB_SERVICE_NAME):/supportpal/config/production" "backup/$(TIMESTAMP)/config/"
```
Comment (bytestream) [inline templates/docker-compose/Makefile:88]:
```suggestion
	@ mkdir -p "backup/$(TIMESTAMP)/config/" && docker cp "$(WEB_SERVICE_NAME):/supportpal/config/production" "backup/$(TIMESTAMP)/config/"
```

File: templates/docker-compose/Makefile [hunk 8]
```diff
@@ -73,3 +74,17 @@ configure:
...
+	mkdir -p "backup/$(TIMESTAMP)/storage/" && docker cp "$(WEB_SERVICE_NAME):/supportpal/storage" "backup/$(TIMESTAMP)/storage/"
```
Comment (bytestream) [inline templates/docker-compose/Makefile:89]:
```suggestion
	@ mkdir -p "backup/$(TIMESTAMP)/storage/" && docker cp "$(WEB_SERVICE_NAME):/supportpal/storage" "backup/$(TIMESTAMP)/storage/"
```

File: templates/docker-compose/Makefile [hunk 9]
```diff
@@ -73,3 +74,17 @@ configure:
...
+	echo "Backups created successfully at $(CURDIR)/backup/$(TIMESTAMP)"
```
Comment (bytestream) [inline templates/docker-compose/Makefile:90]:
```suggestion
	@ echo "Backups created successfully at $(CURDIR)/backup/$(TIMESTAMP)"
```

File: templates/docker-compose/Makefile (hunk 1 above)
Comment (bytestream) [reply templates/docker-compose/Makefile:79]: Perhaps this needs to be clearer that you need to change `APP_VERSION` first...

File: templates/docker-compose/Makefile (hunk 3 above)
Comment (bytestream) [reply templates/docker-compose/Makefile:2]: It creates a timestamp e.g. `1618490824`

**Review by bytestream — COMMENTED** [review]
We should run `backup` as part of `make uninstall` too

File: templates/docker-compose/Makefile (hunk 3 above)
Comment (Gaitholabi) [reply templates/docker-compose/Makefile:2]: The main reason we have this is to eliminate backups overwriting each other. I'll see a friendlier format

Comment (bytestream) [reply templates/docker-compose/Makefile]:
```
$ date +%s-%F
1618495023-2021-04-15
```

Above might be a good format? Easy to sort / manipulate with the timestamp prefix.

Comment (Gaitholabi) [reply templates/docker-compose/Makefile]: don't have a strong opinion on this. I'll use `date +%s-%F`

**Review by bytestream — APPROVED** [review]

**Pull request merged**